File: cpp/1514.path-with-maximum-probability.cpp

#include <queue>
#include <unordered_map>
#include <utility>
#include <vector>

using std::pair;
using std::priority_queue;
using std::unordered_map;
using std::vector;

class Solution {
public:
  double maxProbability(int n, vector<vector<int>> &edges,
                        vector<double> &succProb, int start_node,
                        int end_node) {

    unordered_map<int, vector<pair<int, double>>> adj;

    for (int i = 0; i < edges.size(); i++) {
      adj[edges[i][0]].push_back({edges[i][1], succProb[i]});
      adj[edges[i][1]].push_back({edges[i][0], succProb[i]});
    }

    priority_queue<pair<double, int>, vector<pair<double, int>>, std::less<>>
        pq;

    pq.push({1.0, start_node});

    vector<double> prob(n, 0.0);
    prob[start_node] = 1.0;

    while (pq.size()) {
      auto [nodeProb, node] = pq.top();
      pq.pop();

      if (prob[node] > nodeProb) {
        continue;
      }

      for (auto [child, childProb] : adj[node]) {
        double newProb = nodeProb * childProb;

        if (prob[child] < newProb) {
          prob[child] = newProb;
          pq.push({newProb, child});
        }
      }
    }

    return prob[end_node];
  }
};

File: python/1514.path-with-maximum-probability.py
import heapq
from collections import defaultdict


class Solution:
    def maxProbability(
        self,
        n: int,
        edges: list[list[int]],
        succProb: list[float],
        start_node: int,
        end_node: int,
    ) -> float:
        adj: defaultdict[int, list[tuple[int, float]]] = defaultdict(list)
        for (u, v), p in zip(edges, succProb):
            adj[u].append((v, p))
            adj[v].append((u, p))

        hq = [(-1.0, start_node)]
        prob = [0.0] * n
        prob[start_node] = 1.0

        while hq:
            p, u = heapq.heappop(hq)

            if prob[u] < -p:
                continue

            for v, pp in adj[u]:
                if prob[v] < -p * pp:
                    prob[v] = -p * pp
                    heapq.heappush(hq, (p * pp, v))

        return prob[end_node]

